### DISCUSSION TITLE
Using Custom Schemas

### DISCUSSION THREAD

**Discussion by huntthomp** [supabase-community/supabase-csharp#173] [Q&A] [answered]
How can you fetch data from a custom schemas? I think in the JavaScript and Flutter clients you can do something like this. 
`final data = await supabase.schema('myschema').from('mytable').select();`
I've looked in the documentation and can't find anything.

**Comment by huntthomp** [comment, marked as answer]
I have solved my own problem. I don't think there is a way to specify schemas like in Js, as of right now. But you can set a specific schema you want to target when instantiating your Supabase client.

```
SupabaseOptions options = new Supabase.SupabaseOptions
{
    Schema = "your_schema"
};

Client supabase = new Supabase.Client(url, key, options);
```

You HAVE to set the schema you want to access as an exposed schema in the API settings in your Supabase project

**Reply by ZadokJoshua** [reply]
In addition to @huntthomp's answer, if your Supabase client instance is managed by the Dependency Injection container, you can utilise **keyed services** (introduced in .NET 8) to handle instances for specific schemas. Here is an example:

```csharp
public static IServiceCollection ConfigureSupabase(this IServiceCollection services, IConfiguration configuration)
{
    var supabaseConfig = configuration.GetSection("SupabaseProject");

    // SupabaseOptions for different schemas
    SupabaseOptions privateSchemaOptions = new() { Schema = "private" };
    SupabaseOptions publicSchemaOptions = new() { Schema = "public" };

    // Register the Supabase clients with different schema options
    services.AddKeyedSingleton<Supabase.Client>(
        "private", (sp, key) => new Supabase.Client(
            supabaseConfig["ProjectUrl"]!,
            supabaseConfig["ServiceRoleKey"],
            privateSchemaOptions
        ));

    services.AddKeyedSingleton<Supabase.Client>(
        "public", (sp, key) => new Supabase.Client(
            supabaseConfig["ProjectUrl"]!,
            supabaseConfig["ServiceRoleKey"],
            publicSchemaOptions
         ));

    return services;
}
```

You can use them as follows:

```csharp
public class SuperRepository([FromKeyedServices("public")] Supabase.Client publicSupabaseClient) : ISuperRepository
{
    public async Task Remove(long superId)
    {
        await publicSupabaseClient.From<Super>().Where(s => s.Id == superId).Delete();
    }
}
```

```csharp
public class SecretRepository([FromKeyedServices("private")] Supabase.Client privateSupabaseClient) : ISecretRepository
{
    public async Task Remove(long secretId)
    {
        await privateSupabaseClient.From<Secret>().Where(s => s.Id == secretId).Delete();
    }
}
```

[Learn more about keyed services](https://andrewlock.net/exploring-the-dotnet-8-preview-keyed-services-dependency-injection-support/)

**Comment by Jonydeere1** [comment]
Im starting an Android App on MAUI, now im working on my Backend (Supabase), and i have 2 projects (Professionals, and Common) each one with lot of schemas (forms, surveys, files, public, personal, etc.).

can i set different clients and schemas?

**Comment by 1knueller** [comment]
is there any way or plan to be able to use one client for more than one schema?
the usualy supabase implementations can use sth like repository.schema("MySchema").table("table") wich would be nice
even more nice would be to have the default schema and maybe add a [Schema("MySchema")] attribute to the class we use like currently we have [Table("data_sets")] on the class
